File: protocol-generator/go/src/peer/smoke_test.go
package peer

// smoke_test.go — the S3 two-peer loopback smoke gate (the phase exit criterion).
//
// Two Go peers talk over real loopback TCP through the full §6.5 dispatch chain.
// Scenario 1 (responder = default seed policy) exercises the handshake, 404,
// authority-gated tree get, capability request, and 8-way request_id demux.
// Scenario 2 (responder = --debug-open-grants + --validate) exercises the Core
// Extensibility Boundary: register live-hook, emit hook, §7a echo. 11 checks
// total; the run is GREEN iff all 11 pass.
//
// This is the Go analogue of every sibling peer's S3 smoke (Common-Lisp/Swift:
// 11/11 loopback). The full validate-peer --profile core conformance run is S4.

import (
	"sync"
	"sync/atomic"
	"testing"

	"github.com/entity-core/entity-core-protocol-go/internal/cbor"
)

func fixedSeed(b byte) []byte {
	s := make([]byte, 32)
	for i := range s {
		s[i] = b
	}
	return s
}

// TestSmokeLoopback is the 11-check two-peer loopback gate.
func TestSmokeLoopback(t *testing.T) {
	scenario1(t)
	scenario2(t)
}

func scenario1(t *testing.T) {
	t.Helper()
	responder, err := NewPeer(fixedSeed(0x11))
	if err != nil {
		t.Fatalf("responder bootstrap: %v", err)
	}
	initiator, err := MakeIdentity(fixedSeed(0x22))
	if err != nil {
		t.Fatalf("initiator identity: %v", err)
	}
	ln, err := responder.Listen(0)
	if err != nil {
		t.Fatalf("listen: %v", err)
	}
	defer ln.Close()
	t.Logf("Responder listening on 127.0.0.1:%d (peer %s)", ln.Port(), responder.LocalPeer())

	cc, err := Dial("127.0.0.1", ln.Port())
	if err != nil {
		t.Fatalf("dial: %v", err)
	}
	defer cc.Close()
	if err := cc.Handshake(initiator); err != nil {
		t.Fatalf("handshake: %v", err)
	}
	remote := cc.RemotePeerID()

	// (1) session established (capability minted)
	if _, ok := cc.Capability(); !ok {
		t.Error("[FAIL] session established (capability minted)")
	} else {
		t.Log("[PASS] session established (capability minted)")
	}
	// (2) remote peer_id matches responder
	if remote != responder.LocalPeer() {
		t.Errorf("[FAIL] remote peer_id matches responder (%s != %s)", remote, responder.LocalPeer())
	} else {
		t.Log("[PASS] remote peer_id matches responder")
	}

	ifaceTarget := ResourceTarget("system/handler/system/tree")

	// (3) unregistered path -> 404
	if r, ok := cc.Execute(initiator, "/"+remote+"/does/not/exist", "noop", EmptyParams()); !ok || ResponseStatus(r) != 404 {
		t.Errorf("[FAIL] unregistered path -> 404 (got %d)", ResponseStatus(r))
	} else {
		t.Log("[PASS] unregistered path -> 404")
	}

	// (4) granted tree get -> 200 + (5) returns a system/handler/interface entity
	rget, ok := cc.Execute(initiator, "/"+remote+"/system/tree", "get", EmptyParams(), ifaceTarget)
	if !ok || ResponseStatus(rget) != 200 {
		t.Errorf("[FAIL] granted tree get -> 200 (got %d)", ResponseStatus(rget))
	} else {
		t.Log("[PASS] granted tree get -> 200")
	}
	if res, ok := ResponseResult(rget); !ok || res.Type != "system/handler/interface" {
		t.Errorf("[FAIL] tree get returns a system/handler/interface entity (got %q)", res.Type)
	} else {
		t.Log("[PASS] tree get returns a system/handler/interface entity")
	}

	// (6) capability request -> 200
	reqGrant := grantSpec{handlers: []string{"system/tree"}, resources: []string{"system/type/*"}, operations: []string{"get"}}
	reqParams := mustEntity("system/capability/request", cbor.NewMap(cbor.Entry("grants", grantsCbor(reqGrant))))
	if rcap, ok := cc.Execute(initiator, "/"+remote+"/system/capability", "request", reqParams); !ok || ResponseStatus(rcap) != 200 {
		t.Errorf("[FAIL] capability request -> 200 (got %d)", ResponseStatus(rcap))
	} else {
		t.Log("[PASS] capability request -> 200")
	}

	// (7) 8-way request_id demux (N7, §6.11)
	const n = 8
	var correlated int64
	var wg sync.WaitGroup
	for i := 0; i < n; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			r, ok := cc.Execute(initiator, "/"+remote+"/system/tree", "get", EmptyParams(), ifaceTarget)
			if ok && ResponseStatus(r) == 200 {
				if res, ok := ResponseResult(r); ok && res.Type == "system/handler/interface" {
					atomic.AddInt64(&correlated, 1)
				}
			}
		}()
	}
	wg.Wait()
	if correlated != n {
		t.Errorf("[FAIL] 8 interleaved requests each correlated -> %d/8", correlated)
	} else {
		t.Log("[PASS] 8 interleaved requests each correlated -> 8/8")
	}
}

func scenario2(t *testing.T) {
	t.Helper()
	responder, err := NewPeer(fixedSeed(0x33), WithOpenGrants(), WithConformance())
	if err != nil {
		t.Fatalf("responder bootstrap: %v", err)
	}
	initiator, err := MakeIdentity(fixedSeed(0x44))
	if err != nil {
		t.Fatalf("initiator identity: %v", err)
	}

	var emitCount int64
	responder.Store().RegisterTreeConsumer(func(TreeEvent) { atomic.AddInt64(&emitCount, 1) })

	ln, err := responder.Listen(0)
	if err != nil {
		t.Fatalf("listen: %v", err)
	}
	defer ln.Close()

	cc, err := Dial("127.0.0.1", ln.Port())
	if err != nil {
		t.Fatalf("dial: %v", err)
	}
	defer cc.Close()
	if err := cc.Handshake(initiator); err != nil {
		t.Fatalf("handshake: %v", err)
	}
	remote := cc.RemotePeerID()
	emitBefore := atomic.LoadInt64(&emitCount)

	// (8) handler register -> 200 (live, not 501) + (9) emit hook fired
	manifest := cbor.NewMap(cbor.Entry("name", cbor.Text("demo")), cbor.Entry("operations", emptyMap()))
	req := mustEntity("system/handler/register-request", cbor.NewMap(cbor.Entry("manifest", manifest)))
	rreg, ok := cc.Execute(initiator, "/"+remote+"/system/handler", "register", req, ResourceTarget("system/handler/demo"))
	if !ok || ResponseStatus(rreg) != 200 {
		t.Errorf("[FAIL] handler register -> 200 (live, not 501) (got %d)", ResponseStatus(rreg))
	} else {
		t.Log("[PASS] handler register -> 200 (live, not 501)")
	}
	if atomic.LoadInt64(&emitCount) <= emitBefore {
		t.Error("[FAIL] emit hook fired on register's tree writes (§6.13(c))")
	} else {
		t.Log("[PASS] emit hook fired on register's tree writes (§6.13(c))")
	}

	// (10) §7a echo -> 200 + (11) returns params verbatim
	payload := mustEntity("primitive/any", cbor.NewMap(cbor.Entry("ping", cbor.Uint(42))))
	recho, ok := cc.Execute(initiator, "/"+remote+"/system/validate/echo", "echo", payload)
	if !ok || ResponseStatus(recho) != 200 {
		t.Errorf("[FAIL] §7a echo -> 200 (got %d)", ResponseStatus(recho))
	} else {
		t.Log("[PASS] §7a echo -> 200")
	}
	if res, ok := ResponseResult(recho); !ok || res.Type != "primitive/any" {
		t.Errorf("[FAIL] §7a echo returns params verbatim (got %q)", res.Type)
	} else {
		t.Log("[PASS] §7a echo returns params verbatim")
	}
}

File: protocol-generator/php/tests/SmokeTest.php
<?php

declare(strict_types=1);

namespace EntityCore\Tests;

use EntityCore\ByteString;
use EntityCore\Ecf;
use EntityCore\Entity;
use EntityCore\Peer;
use EntityCore\Session;
use EntityCore\Transport;
use EntityCore\Wire;
use PHPUnit\Framework\Attributes\Group;
use PHPUnit\Framework\TestCase;

/**
 * S3 two-peer loopback smoke test (the phase exit gate; 11/11).
 *
 * Two PHP peers talk over real loopback TCP through the full §6.5 dispatch chain,
 * multiplexed on ONE single-thread {@see \EntityCore\EventLoop} (A-PHP-005): a
 * RESPONDER listens; an INITIATOR (a second identity) dials it and drives the §4.1
 * forward handshake (hello → authenticate), then:
 *  - 404 on an unregistered path (no handler resolved);
 *  - an authority-gated tree get (200) over the §4.4 discovery floor;
 *  - a capability request (200);
 *  - 8-way request_id demux of concurrently-issued replies (N7, §6.11) — 8
 *    requests in flight on the one connection, resolved out of order by the loop.
 *
 * A second scenario exercises the v7.74 Core Extensibility Boundary
 * (--debug-open-grants + --validate): the register live-hook (§6.13(a)), the emit
 * hook firing on register's tree writes (§6.13(c)), the §7a echo handler, AND the
 * §6.11 dispatch-outbound reentry (the validator-as-B surface S4 needs).
 *
 * The full validate-peer --profile core run is S4. This smoke proves the
 * wire-level peer surface so S4 can run the oracle.
 */
#[Group('loopback')]
final class SmokeTest extends TestCase
{
    /** @var list<bool> */
    private array $results = [];

    private function check(string $name, bool $ok): bool
    {
        $this->results[] = $ok;
        \fwrite(\STDOUT, '  [' . ($ok ? 'PASS' : 'FAIL') . "] {$name}\n");
        return $ok;
    }

    private function seed(int $b): string
    {
        return \str_repeat(\chr($b), 32);
    }

    public function testTwoPeerLoopback(): void
    {
        $this->runCoreScenario();
        $this->runExtensibilityScenario();
        $pass = \count(\array_filter($this->results));
        $total = \count($this->results);
        \fwrite(\STDOUT, "\nSMOKE: " . ($pass === $total ? 'PASS' : 'FAIL') . " ({$pass}/{$total})\n");
        self::assertSame($total, $pass, 'two-peer loopback must be all-PASS');
    }

    // ── Scenario 1: core ops (responder = default seed policy) ──────────────────────

    private function runCoreScenario(): void
    {
        $transport = Transport::withLoop();
        $responder = Peer::create($this->seed(0x11));
        [$listener, $port] = $transport->startListener($responder, 0);
        try {
            $initiator = Peer::create($this->seed(0x22));
            $s = $transport->dial($initiator, '127.0.0.1', $port);
            try {
                $remote = $s->remotePeerId;
                $this->check('session established (capability minted)', $s->capability !== null);
                $this->check('remote peer_id matches responder', $remote === $responder->localPeer);

                // 404 on an unregistered path
                $r404 = $s->execute("/{$remote}/does/not/exist", 'noop', Wire::emptyParams(), null);
                $this->check('unregistered path -> 404', $r404 !== null && Wire::responseStatus($r404) === 404);

                // authority-gated tree get (200) over the discovery floor.
                $ifaceTarget = Wire::resourceTarget('system/handler/system/tree');
                $rget = $s->execute("/{$remote}/system/tree", 'get', Wire::emptyParams(), $ifaceTarget);
                $this->check('granted tree get -> 200', $rget !== null && Wire::responseStatus($rget) === 200);
                $res = $rget !== null ? Wire::responseResult($rget) : null;
                $this->check('tree get returns a system/handler/interface entity',
                    $res !== null && $res->type === 'system/handler/interface');

                // capability request (200)
                $reqGrant = Peer::grant(['system/tree'], ['system/type/*'], ['get'], null);
                $reqParams = Entity::make('system/capability/request', Ecf::map('grants', [$reqGrant]));
                $rcap = $s->execute("/{$remote}/system/capability", 'request', $reqParams, null);
                $this->check('capability request -> 200', $rcap !== null && Wire::responseStatus($rcap) === 200);

                // 8-way request_id demux (N7, §6.11) — 8 requests in flight at once.
                $rids = [];
                for ($i = 0; $i < 8; $i++) {
                    $rids[] = $s->executeAsync("/{$remote}/system/tree", 'get', Wire::emptyParams(),
                        Wire::resourceTarget('system/handler/system/tree'));
                }
                $replies = $s->awaitAll($rids);
                $correlated = 0;
                foreach ($rids as $rid) {
                    $r = $replies[$rid] ?? null;
                    if ($r === null || Wire::responseStatus($r) !== 200) {
                        continue;
                    }
                    $rr = Wire::responseResult($r);
                    // the reply's request_id MUST equal the one we sent (demux correctness)
                    if ($rr !== null && $rr->type === 'system/handler/interface'
                        && ($r->root->text('request_id') === $rid)) {
                        $correlated++;
                    }
                }
                $this->check("8 interleaved requests each correlated -> {$correlated}/8", $correlated === 8);
            } finally {
                $s->close();
            }
        } finally {
            $listener->close();
        }
    }

    // ── Scenario 2: the v7.74 Core Extensibility Boundary over the wire ─────────────

    private function runExtensibilityScenario(): void
    {
        $transport = Transport::withLoop();
        $responder = Peer::create($this->seed(0x33), openGrants: true, conformance: true);
        $emitEvents = 0;
        $responder->store->registerTreeConsumer(function () use (&$emitEvents): void {
            $emitEvents++;
        });
        [$listener, $port] = $transport->startListener($responder, 0);
        try {
            // the initiator is ALSO a --validate peer so the §6.11 reentry echo
            // round-trips (B originates echo back to A; A must serve it).
            $initiator = Peer::create($this->seed(0x44), openGrants: true, conformance: true);
            $s = $transport->dial($initiator, '127.0.0.1', $port);
            try {
                $remote = $s->remotePeerId;
                $emitBefore = $emitEvents;

                // register live-hook (§6.13(a))
                $manifest = Ecf::map('name', 'demo', 'operations', Ecf::emptyMap());
                $req = Entity::make('system/handler/register-request', Ecf::map('manifest', $manifest));
                $rreg = $s->execute("/{$remote}/system/handler", 'register', $req,
                    Wire::resourceTarget('system/handler/demo'));
                $this->check('handler register -> 200 (live, not 501)',
                    $rreg !== null && Wire::responseStatus($rreg) === 200);
                $this->check('emit hook fired on register tree writes (§6.13(c))', $emitEvents > $emitBefore);

                // §7a echo conformance handler (resolve→dispatch)
                $payload = Entity::make('primitive/any', Ecf::map('ping', 42));
                $recho = $s->execute("/{$remote}/system/validate/echo", 'echo', $payload, null);
                $this->check('§7a echo -> 200', $recho !== null && Wire::responseStatus($recho) === 200);
                $res = $recho !== null ? Wire::responseResult($recho) : null;
                $this->check('§7a echo returns params verbatim', $res !== null && $res->type === 'primitive/any');

                // §6.11 dispatch-outbound REENTRY: B originates an outbound EXECUTE
                // back over THIS inbound connection to A; A's reader dispatches it.
                $this->check('§6.11 dispatch-outbound reentry round-trips (B→A echo over inbound conn)',
                    $this->runReentryProbe($s, $remote));
            } finally {
                $s->close();
            }
        } finally {
            $listener->close();
        }
    }

    private function runReentryProbe(Session $s, string $remote): bool
    {
        $params = Entity::make('primitive/any', Ecf::map(
            'target', 'system/validate/echo',
            'operation', 'echo',
            'value', Ecf::map('ping', 7),
            'reentry_capability', $s->capability->toCbor(),
            'reentry_granter', $s->granterPeer->toCbor(),
            'reentry_cap_signature', $s->capSignature->toCbor(),
        ));
        $r = $s->execute("/{$remote}/system/validate/dispatch-outbound", 'dispatch', $params, null);
        if ($r === null) {
            return false;
        }
        // Outer 200 = the §6.11 reentry round-tripped end-to-end: B's
        // dispatch-outbound handler originated an EXECUTE back to A over the SAME
        // inbound connection, A's reader dispatched it and replied, and B
        // correlated the reply by request_id and returned it. The INNER status
        // reflects A's authz verdict on the reentrant EXECUTE (a §5.2 cap check);
        // S4's validator supplies the cross-peer reentry cap that makes it 200.
        $inner = Wire::responseResult($r)?->uint('status');
        \fwrite(\STDOUT, '    (reentry round-tripped; inner echo verdict=' . ($inner !== null ? \gmp_strval($inner) : 'null') . ")\n");
        return Wire::responseStatus($r) === 200;
    }
}
